src/refactorings/convert-switch-to-if-else/convert-switch-to-if-else.ts

```diff
@@ -0,0 +1,172 @@
+import { Editor, Code, ErrorReason } from "../../editor/editor";
+import { Selection } from "../../editor/selection";
+import * as t from "../../ast";
+import { last, allButLast } from "../../array-helpers";
+
+export { convertSwitchToIfElse, createVisitor as hasSwitchToConvert };
+
+async function convertSwitchToIfElse(
+  code: Code,
+  selection: Selection,
+  editor: Editor
+) {
+  const updatedCode = updateCode(t.parse(code), selection);
+
+  if (!updatedCode.hasCodeChanged) {
+    editor.showError(ErrorReason.DidNotFindSwitchToConvert);
+    return;
+  }
+
+  await editor.write(updatedCode.code);
+}
+
+function updateCode(ast: t.AST, selection: Selection): t.Transformed {
+  return t.transformAST(
+    ast,
+    createVisitor(selection, (path, convertedNode) => {
+      path.replaceWith(convertedNode);
+      path.stop();
+    })
+  );
+}
+
+function createVisitor(
+  selection: Selection,
+  onMatch: (
+    path: t.NodePath<t.SwitchStatement>,
+    convertedNode: t.IfStatement | t.SwitchStatement
+  ) => void
+): t.Visitor {
+  return {
+    SwitchStatement(path) {
+      if (!selection.isInsidePath(path)) return;
+
+      // Since we visit nodes from parent to children, first check
+      // if a child would match the selection closer.
+      if (hasChildWhichMatchesSelection(path, selection)) return;
+
+      const convertedNode = convert(path.node);
+      if (convertedNode === path.node) return;
+
+      onMatch(path, convertedNode);
+    }
+  };
+}
+
+function hasChildWhichMatchesSelection(
+  path: t.NodePath,
+  selection: Selection
+): boolean {
+  let result = false;
+
+  path.traverse({
+    SwitchStatement(childPath) {
+      if (!selection.isInsidePath(childPath)) return;
+
+      const convertedNode = convert(childPath.node);
+      if (convertedNode === childPath.node) return;
+
+      result = true;
+      childPath.stop();
+    }
+  });
+
+  return result;
+}
+
+function convert(node: t.SwitchStatement): t.SwitchStatement | t.IfStatement {
+  try {
+    return convertNode(node);
+  } catch (err) {
+    console.log(err);
```

Comment:
We don't want to `console.log()` anything generally. If there's an error, it should be explicit to the user.

But we don't want such error to be explicit. Nor we want to log it: this refactoring will be checked anytime the cursor is moved (so we can propose the quick fix to the user if it can be performed). So that would log an error each time we move the cursor in a switch statement that can't be converted.

I think we mostly use the error to control the flow here (that is, to return an `IfStatement` instead of the `SwitchStatement`). 

I think we could use the NullObject pattern instead. It will probably simplify this.

Reply:
Ahah I commit the console.log by mistake it was only there for debugging purposes... Good catch